src/components/sketches/Scene5_happy.ts

```diff
@@ -0,0 +1,290 @@
+    p.angleMode(p.DEGREES);
+    p.background(0);
+    // Sets 0,0 to centre
+    p.translate(p.width / 2, p.height / 2);
+    // Max number of ringParticles
+    p.frameRate(30);
+  };
+
+  p.windowResized = () => {
+    p.resizeCanvas(window.innerWidth, window.innerHeight, true);
+  };
+
+  p.draw = () => {
+    p.background(0, 0, 0, 25); // easy trails but deletes rings quickly
+
+    p.ellipse(p.mouseX,p.mouseY, 5,5);
+    if (p.random(1) < 0.03) {
+      fireworks.push(new Firework());
+    }
+
+    for (let i = fireworks.length - 1; i >= 0; i--) {
+      fireworks[i].update();
+      fireworks[i].show();
+
+      if (fireworks[i].done()) {
+        fireworks.splice(i, 1);
+      }
+    }
+
+    // If too many ringParticles on screen
+    if (ringParticles.length > MAX_RING_PARTICLES) {
+      // Delete old ringParticles (from beginning of array)
+      ringParticles.splice(0, p.floor((ringParticles.length - MAX_RING_PARTICLES * 0.8) * EASING));
+    }
+
+    // Loop through the array and show each particle
+    for (let i = 0; i < ringParticles.length; i++) {
+      ringParticles[i].update();
+      if (ringParticles[i].delete()) {
+        ringParticles.splice(i, 1);
+      } else {
+        ringParticles[i].show();
+      }
+    }
+  };
+  // Spawn ringParticles in ring with centre mouse x and y coords
+  p.mouseClicked = () => {
+    for (let i = 0; i < RING_COVERAGE; i += 4) {
+      const c = new ringParticle(p.mouseX, p.mouseY, i);
+      ringParticles.push(c);
+    }
+  };
+
+  class Firework {
+    explodeSize: number;
+    firework: Particle;
+    exploded: boolean;
+    particles: Array<Particle>;
+
+    constructor() {
+      this.explodeSize = p.random(0, EXPLODE_RANGE);
+      this.firework = new Particle(p.random(p.width), p.height, true);
+      this.exploded = false;
+      this.particles = [];
+    }
+
+    update() {
+      if (!this.exploded) {
+        this.firework.applyForce(gravity);
+        this.firework.update();
+
+        if (this.firework.vel.y >= 0) {
+          this.exploded = true;
+          this.explode();
+        }
+      }
+
+      for (let i = this.particles.length - 1; i >= 0; i--) {
+        this.particles[i].applyForce(gravity);
+        this.particles[i].update();
+
+        if (this.particles[i].done()) {
+          this.particles.splice(i, 1);
+        }
+      }
+    }
+
+    done() {
+      if (this.exploded && this.particles.length === 0) {
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    show() {
+      if (!this.exploded) {
+        this.firework.show();
+      }
+
+      for (let i = 0; i < this.particles.length; i++) {
+        this.particles[i].show();
+      }
+    }
+
+    explode() {
+      for (let i = 0; i < this.explodeSize * EXPLODE_OFFSET; i++) {
+        const particle = new Particle(this.firework.pos.x, this.firework.pos.y, false);
+        this.particles.push(particle);
+      }
+    }
```

Comment:
since `this.explode()` should only happen once, we should just assign `this.particles` to a new array with map!

I prefer maps because they are more *javascript-ey* and after reading this [article](https://medium.com/@ExplosionPills/map-vs-for-loop-2b4ce659fb03) they are also faster!

```suggestion
      this.particles = Array(this.explodeSize*EXPLODE_OFFSET).fill(0).map(() => 
         new Particle(this.firework.pos.x, this.firework.pos.y, false));
    }
```

Reply:
nice!

Reply:
i just tried this actually and it broke it :(

Reply:
the syntax might be wrong haha, maybe look into that!

Reply:
o this also has `this.firework` so maybe do `this._particle`

Reply:
`this._particles = Array(this.explodeSize*EXPLODE_OFFSET).fill(0).map(() => 
         new Particle(this._firingParticle.pos.x, this._firingParticle.pos.y, false));`
         
this is what i changed it to, but could it be because the map creates a new array and we want it push particles into the array. so map overwrites the particles array while old implementation appends so it just updates. I read through the article and thats kinda what i found.